[quant][graphmode] Whiltelist and fusion support for quantized::linear - matmul(with bias)

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26204 [quant][graphmode] Whiltelist and fusion support for quantized::linear - matmul(with bias)**
* #26209 [quant][graphmode] Whiltelist and fusion support for quantized::linear - matmul(without bias)
* #26208 [quant][graphmode] Whiltelist and fusion support for quantized::linear - addmm

Summary:
Support quant fusion for `matmul` with bias to `quantized::linear`.

Test Plan:
python test/test_jit.py 'TestJit.test_quant_fusion'

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/26204

Differential Revision: [D17380073](https://our.internmc.facebook.com/intern/diff/D17380073)

jerryzh168 committed:
[quant][graphmode] Whiltelist and fusion support for quantized::linear - Part 2

Summary:
Support quant fusion for `matmul` with bias.

Test Plan:
python test/test_jit.py 'TestJit.test_quant_fusion'

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Whiltelist and fusion support for quantized::linear - Part 2"


Summary:
Support quant fusion for `matmul` with bias.

Test Plan:
python test/test_jit.py 'TestJit.test_quant_fusion'

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Whiltelist and fusion support for quantized::linear - Part 2"


Summary:
Support quant fusion for `matmul` with bias.

Test Plan:
python test/test_jit.py 'TestJit.test_quant_fusion'

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -691,6 +698,34 @@ graph(%a_quant, %w_quant, %b_quant, %a_scale, %a_zero_point, %a_dtype, %w_scale,
+       // matmul(with bias) -> quantized::linear
+       {R"(
+graph(%a_quant, %w_quant, %b_quant, %a_scale, %a_zero_point, %a_dtype, %w_scale, %w_zero_point, %w_dtype, %b_scale, %b_zero_point, %b_dtype, %r_scale, %r_zero_point, %r_dtype, %4):
+        %a_intrepr = aten::int_repr(%a_quant)
```
Looks like we're using the same replacement pattern for several transforms. How about hoisting it to a separate variable, like
```
std::string fbgemm_linear_pattern = R"(
graph(%a_quant, %w_quant, %a_scale, %a_zero_point, %a_dtype, %w_scale, %w_zero_point, %w_dtype, %r_scale, %r_zero_point, %r_dtype):
        %0 : int = prim::Constant[value=0]()
        %1 : int = prim::Constant[value=1]()
        %2 : int = prim::Constant[value=2]()
        %3 : int = prim::Constant[value=3]()
        %in_param : int[] = prim::ListConstruct(%0, %2, %3, %1)
        %a_perm : Tensor = aten::permute(%a_quant, %in_param)
        %w_quant_t = aten::t(%w_quant)
        %w_perm : Tensor = aten::permute(%w_quant_t, %in_param)
        %w_packed = quantized::fbgemm_linear_prepack(%w_perm)
        %bias: Tensor? = prim::Constant()
        %r = quantized::fbgemm_linear(%a_perm, %w_packed, %bias, %r_scale, %r_zero_point)
        %out_param : int[] = prim::ListConstruct(%0, %3, %1, %2)
        %r_perm = aten::permute(%r, %out_param)";
```
and then rewrite the patterns as:
```
// matmul(with bias) -> quantized::linear
       {<<<pattern>>>,
        fbgemm_linear_pattern},
```

jerryzh168 replied:
OK, let me check, I'll combine them if they are the same

ZolotukhinM reviewed (approved):
Looks good!

jerryzh168 committed:
Update on "[quant][graphmode] Whiltelist and fusion support for quantized::linear - matmul(with bias)"


Summary:
Support quant fusion for `matmul` with bias to `quantized::linear`.

Test Plan:
python test/test_jit.py 'TestJit.test_quant_fusion'

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/26204

jerryzh168 committed:
Update on "[quant][graphmode] Whiltelist and fusion support for quantized::linear - matmul(with bias)"


Summary:
Support quant fusion for `matmul` with bias to `quantized::linear`.

Test Plan:
python test/test_jit.py 'TestJit.test_quant_fusion'

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/26204
